core/sql/optimizer/costmethod.cpp

```diff
@@ -13997,15 +13997,188 @@ CostMethodFastExtract::computeOperatorCostInternal(RelExpr* op,
+  //
+  // ************************************************************
+
+  // ---------------------------------------------------------------------
+  // Synthesize the cost vectors.
+  // ---------------------------------------------------------------------
+  SimpleCostVector cvFR;
+  SimpleCostVector cvLR;
+
+  // For now, we don't bother to estimate CPU time, I/O time, transfer 
+  // time or idle time, since we really are only supporting the new 
+  // cost model.
+  //
+  // Future possible improvements:
+  //
+  // 1. Take into account HBase memstore insertion cost. The memstore
+  // uses a Red-Black tree which has o(n * log(n)) insertion cost. To
+  // model this correctly, we'd need to take into account the number
+  // of HBase regions rather than the number of ESPs, that is, to
+  // divide the number of probes by the number of HBase regions to find
+  // n. This cost will be paid no matter how many inserting streams
+  // there are so by itself this may not be interesting. It would only
+  // be interesting if there were a choice in the plan between inserting
+  // and not inserting (e.g. if we were considering bypassing the
+  // memstore, or if we were considering storing an intermediate result,
+  // neither of which are choices we examine today).
+  //
+  // 2. Take into account whether the probes are in key order. There
+  // is anecdotal evidence that if the probes are in key order, then
+  // memstore insertion cost is less. Possibly this is true only if
+  // inserting at the end or the beginning of the key range in a 
+  // partition; intuitively inserting in the middle would seem to incur
+  // the full insertion cost. This is worthwhile taking into 
+  // consideration as it opens the possibility of choosing between a
+  // plan that sorts rows in Trafodion before passing them to HBase
+  // vs. a plan that does not. To make this calculation we must know
+  // the memstore insertion cost (point 1 above), the order of the 
+  // probes, whether the ESPs are aligned to the Regions of the 
+  // target table, and whether we are inserting at the beginning or
+  // end of the key range. A first approximation to the last item
+  // would be whether the target table is empty. This is interesting
+  // because the case that we are doing an INSERT/SELECT into a new
+  // table is likely to be common.
+  //
+  // 3. Take into account memstore flush cost. We could add I/O time
+  // for flushes. For example, we could compare the number of probes
+  // per HBase Region with the number of rows that would cause a
+  // flush (the latter can be obtained from 
+  // HbaseClient::estimateMemStoreRows() and is a function of the
+  // HBase parameter hbase.hregion.memstore.flush.size). Again, this
+  // cost will be paid no matter the plan choice so this is not
+  // interesting today. As with point 1, it becomes interesting only
+  // if there is a plan choice between inserting via memstore or not.
+  //
```

Comment:
Yes, this will be another future improvement.  Basically, if the # of rows to be inserted is over the threshold of flush,  we include the cost of flush. The cost of insertion into memory store will be two parts: nlogn + mlogm, where n is the number of rows that will lead to the flush.  m is the # of rows that would be kept in the memstore after the flush.